### PR TITLE
Go (surrounding game) demo

### PR DESCRIPTION
**Pull request by gcentauri**
I borrowed the css/svg for the board from here https://blog.appsignal.com/2019/06/18/elixir-alchemy-building-go-with-phoenix-live-view.html which is something @andrewek sent me about a year ago.

Once I figured out how to get the assets to load, getting stone placement to work with Motion was super straight forward! Very nice!

go

This very naively broadcasts the same board to anyone connected to the server using a class variable for shared state. Eventually I'd like to update it so it can save game records and players could reconnect to a game given its id.

### CODE REVIEW

**Comment by gcentauri** [comment]
@latortuga - I didn't have much time, so I just did a very naive implementation of broadcasting board state, but it seems to work? of course, this implementation will only allow a single game at a time on the server xD. Given more time I'd do something similar to the clicker game to allow players to join or just observe.

**Comment by gcentauri** [comment]
Still using a class variable for shared state, and it might be a horrible implementation... but you can start independent games and send a link to a friend to play. I implemented "self capture" rather than an illegal move, because its quick for now, and still Ko is incomplete.

I'm not sure how I'd want to store games in a DB yet, so I just used the class variable for "temporary" persistence. Ofc, this could cause a problem in memory if too many games get created. Obviously more work could go into this.

File: app/assets/stylesheets/go.scss.erb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        padding-bottom: 100%;
+        width: 100%;
+        margin: 1em 0;
+    }
+
+    button{
+        font-size: 13px;
+    }
+
+    .board button{
```
Comment (Gmfholley) [inline app/assets/stylesheets/go.scss.erb:17]: Since we're using sass, let's nest the selectors for clarity.  I think that means that this can go inside the `.board` section abover.

File: app/components/go_component.html.erb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    <div>
+      <% @captures["white"].times do %>
+      <span class="white"></span>
+      <% end %>
+    </div>
+  </div>
+
+  <div class="board <%= @current %>">
+    <% @board.each_with_index do |v,i| %>
+    <%= button_tag "", class: v.to_s, data: { motion: "place", index: i.to_s } %>
```
Comment (Gmfholley) [inline app/components/go_component.html.erb:17]: Just a note that since this gets converted into HTML, the numbers are automatically converted into strings, and the nil board vertices are empty as well.

File: app/components/go_component.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    @board = @game.display
+    @current = @game.current
+    @captures = @game.captures
+
+    stream_from "go:#{@key}", :player_move
+  end
+
+  def place(event)
+    index = event.target.data[:index].to_i
+    pos = GoGame::Pos.new((index / 9), index.modulo(9))
```
Comment (Gmfholley) [inline app/components/go_component.rb:20]: My understanding is that this is hard-coding in a game size of 9.

Should we add a comment here, or check the size from the game instead?

File: app/components/go_component.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    if @game.legal_move?(pos)
+      @game.place(pos)
+      @game.next_player
+      GoGame.update(key: @key, game: @game)
```
Comment (Gmfholley) [inline app/components/go_component.rb:25]: I like how clearly this reads.

File: app/components/go_component.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    end
+  end
+
+  def reset
```
Comment (Gmfholley) [inline app/components/go_component.rb:29]: I think this functionality isn't built out.

One reason we might not want this yet is that the url will stay the same but the game will be different, so a player wanting to invite another player to a new game will not be able to (I think)?

File: app/components/go_component.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    @board = @game.display
+    @current = @game.current
+    @captures = @game.captures
+    broadcast_board
+  end
+
+  private
+
+  def player_move(message)
+    @game = GoGame.find(key: @key)
```
Comment (Gmfholley) [inline app/components/go_component.rb:40]: So I played with this a little bit.  There is so much going on with the internal state of the game!

The game doesn't work properly unless you both find the game and use the broadcast state for the board (at least).

I am concerned that what this shows is that there are two sources of truth:
a) the broadcast state of the board
b) the "saved" record of the game stored in the class variable

This might indeed be solved by storing this in a db.  But I wanted to point out that we'd probably prefer only one source of truth going forward.

File: app/models/go_game.rb [hunk 7]
```diff
@@ -0,0 +1,143 @@
+# this class represents everything needed to build a square Go board,
+# track the state of the game, and possibly produce a game record based
+# on the stones placed on the board.
+class GoGame
```
Comment (Gmfholley) [inline app/models/go_game.rb:4]: Wow!  There is so much going on here.  Congrats on making it all work.

Despite remembering your talk, I am not very familiar with the rules of Go, but I tested capturing, next player moves, and legal and illegal moves.

Your game handles them all really well! 🚀 

I did find some of this code hard to follow, specifically the code updating the groups.  

But I _love_ the Structs in here helping you save intermediate state.
I really like all the comments.
You've named things particularly well.

If you wanted to refactor, I might consider:
* making the structs their own file
* creating a service object to handle the logic inside the update_groups piece

File: app/controllers/go_controller.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+  private
+
+  def key
+    @key ||= params[:id]
+  end
+
+  def enforce_valid_game_key
+    unless key =~ /\A[A-Za-z\d]{10}\z/
+      redirect_to clicker_games_path, error: "Invalid key" and return false
```
[the remark this answers is not in the document]
Comment (gcentauri) [reply app/controllers/go_controller.rb:26]: oops, yeah thanks for finding the copy-pasta :)

File: app/assets/config/manifest.js [hunk 9]
```diff
@@ -1,2 +1,5 @@
 //= link_tree ../images
 //= link_directory ../stylesheets .css
+//= link grid.svg
+//= link application.scss
+//= link go.scss.erb
```
[the remark this answers is not in the document]
Comment (gcentauri) [reply app/assets/config/manifest.js:5]: no, i just didn't understand how the pipeline works best. pretty new to that part of rails/webpack. i was just having trouble getting the css to work and tried several things.

File: app/components/go_component.rb (hunk 3 above)
Comment (gcentauri) [reply app/components/go_component.rb:20]: yeah i'll add a comment. i don't have the assets yet for making the larger size boards. probably there will be a button to determine the size. i was thinking i might want to make a public method on the game for doing this instead of directly building the new position struct here in the component anyway, so yes, the game object can determine the correct position based on the board size.

File: app/components/go_component.rb (hunk 5 above)
Comment (gcentauri) [reply app/components/go_component.rb:29]: yeah, this is vestigial. i forgot to remove this after moving logic to support multiple games with ids instead of just one.

File: app/components/go_component.rb (hunk 6 above)
Comment (gcentauri) [reply app/components/go_component.rb:40]: yeah i was a bit concerned about this. for some reason, if i didn't update the object in the `@game` variable in the component, a new object would get saved in the class variable. I'd hoped that if all components referred to the same object, i could achieve the "shared state". though this is probably "dangerous" because multiple clients could get in a bad state if they all try to make changes.

so this was the "solution" that seemed to work. updating the local variable to "re-fetch" the object stored in the class seemed to work.  probably i should update the component pieces then based on the game object and not use the broadcast message to send any state at all. just broadast something like "updated". 

i suppose this same thing would need to be used with the DB to ensure a component has an updated version of the record? I'm not sure how AR works if there's an object in memory that gets updated.

File: app/models/go_game.rb (hunk 7 above)
Comment (gcentauri) [reply app/models/go_game.rb:4]: i will work on cleaning it up more when i get time. i really appreciate you taking time to do a review @Gmfholley thank you so much!

**Comment by gcentauri** [comment]
@Gmfholley - took your thoughtful suggestions and tried to implement them. probably still some conventions that could be cleaned up (idk if everything belongs in the `models` directory)

i think the work i did may have cleared up some of the state and "source of truth" issues as well. i'm still slightly uncomfortable with it, and there should probably be some more separation of concerns (game state, game logic, game display) but its getting there.

i wondered about how much of an issue it would be to keep the games alive, i think i could add in something like drew has to delete them once the last player disconnects? i should look at the clicker code and see if i can get that in, and then i think this might be good to go for a beta test?

File: app/controllers/go_controller.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  helper_method :key
+
+  layout "go"
+
+  def create
+    game = Go::Game.create
+
+    # TODO: Games persist in memory for 1 hour, remove when shift to DB persistence
+    GoGameCleanupJob.set(wait: 1.hour).perform_later(key: game.key)
```
Comment (gcentauri) [inline app/controllers/go_controller.rb:12]: this might be a silly idea, we don't have any job backend (but probably don't need one for this little demo site) but I wanted to have something that could clean up the memory just in case. was thinking about adding a clock to say how much time is left using the periodic timer, but probably more trouble than its worth. maybe i'll add something to the view that's like "this game will expire at X:XX UTC"

File: test/models/go_game_test.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
+require "test_helper"
+
+describe GoGame do
+  let(:subject) { GoGame.new(key: "foo") }
```
Comment (gcentauri) [inline test/models/go_game_test.rb:4]: this all needs to be rewritten

**Comment by gcentauri** [comment]
@Gmfholley @latortuga @alecdotninja 

I feel confident in merging this in now. The code could be better, but I think everything works. Multiple connections, game syncing, illegal moves, capture, etc. when all components are disconnected, the game instance in memory should be deleted. there may still be some memory issues with storing games in the class variable, but i doubt anything too serious.
